octobot_trading/exchanges/util/parser/orders_parser.py

```diff
@@ -0,0 +1,743 @@
+                    check_completeness=check_completeness,
+                )
+                for raw_order in raw_orders
+            ]
+            if self.raw_records
+            else []
+        )
+        if check_completeness:
+            self._create_debugging_report()
+        return self.formatted_records
+
+    async def parse_order(
+        self,
+        raw_order: dict,
+        order_type: str = None,
+        quantity: decimal.Decimal = None,
+        price: decimal.Decimal = None,
+        status: str = None,
+        symbol: str = None,
+        side: str = None,
+        timestamp: int or float = None,
+        check_completeness: bool = True,
+    ) -> dict:
+        """
+        use this method to format a single order
+
+        :param raw_order: raw order with eventually missing data
+
+        optional:
+        :param status: to use if it's missing in the order
+        :param order_type: to use if it's missing in the order
+        :param price: to use if it's missing in the order
+        :param quantity: to use if it's missing in the order
+        :param symbol: to use if it's missing in the order
+        :param side: to use if it's missing in the order
+        :param timestamp: to use if it's missing in the order
+
+        :param check_completeness: if true checks all attributes, if somethings missing it'll try to fetch it from the exchange
+
+        :return: formatted order dict (100% complete or we raise NotImplemented)
+
+        """
+        check_completeness = (
+            check_completeness if check_completeness is not None else True
+        )
+        # todo post only
+        self.fetched_order = None  # clear previous fetched order
+        self._ensure_dict(raw_order)
+        self._parse_id()
+        self._parse_status(status)
+        self._parse_timestamp(timestamp)
+        self._parse_symbol(symbol)
+        self._parse_side(side)
+        self._parse_type(order_type)  # parse after side
+        self._parse_taker_or_maker()  # parse after type
+        self._parse_price(price)  # parse after type, status
+        self._parse_filled_price()  # parse after price
+        self._parse_average_price()
+        self._parse_amount(quantity)
+        self._parse_remaining()  # parse after amount and status
+        self._parse_filled_amount()  # parse after amount and remaining
+        self._parse_cost()
+        self._parse_reduce_only()  # parse after type
+        self._parse_tag()
+        self._parse_fees()
+        # self._parse_datetime(timestamp)  # remove? is it used?
+        # self._parse_last_trade_timestamp()  # remove? is it used?
+        # self._parse_quantity_currency()  # remove? is it used?
+
+        if check_completeness:
+            await self._fetch_if_missing()
+            self._create_debugging_report_for_record()
+
+        return self.formatted_record
+
+    def _parse_status(self, missing_status_value):
+        def missing_status():
```

Comment:
we should avoid closures as much as possible, especially if we cythonize this code.
I made a benchmark to measure the performance impact of closures here: https://github.com/Drakkar-Software/OctoBot-Benchmarks/pull/22

Reply:
crap then I'll change that

Reply:
I added some benchmarks for class methods as well